Eth signing: Add tx hash to SigningOutput

catenocrypt opened the pull request:
## Description

Extend Ethereum SigninOutput with tx hashes: the tx pre-hash (used for singing), and the final hash of the signed tx (which will be the ID of this tx).  This can be useful in some use cases (e.g. own signing).
Eth proto file is affected.

## Testing instructions

Unit tests, `*Ethereum*`.

## Types of changes

<!-- * Bug fix (non-breaking change which fixes an issue) -->

* New feature (non-breaking change which adds functionality)

<!-- * Breaking change (fix or feature that would cause existing functionality to change) -->

## Checklist

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

hewigovens commented:
this is not a clean branch?  it will break sdk users again, it's simple as `Hash.keccak(data)` do we really need it?

vikmeup commented:
> this is not a clean branch? it will break sdk users again, it's simple as `Hash.keccak(data)` do we really need it?

That's a good point, unless we append new properties to protobuf (instead of putting them in front), then it's fine.

catenocrypt commented:
> this is not a clean branch? it will break sdk users again, it's simple as `Hash.keccak(data)` do we really need it?

The point is the pre-hash field, that is hard to compute, message encoding is needed for that.  Motivation was that recent request for RSV signing, pre-hash would allow custom signing.  It may be useful in other cases as well.
The post-signing hash is indeed just a hash on the `encoded` field, I removed it.
It affects proto file, protobuf binary format, so yes, sdk users as well.  I changed so that new field is appended, so it should be backwards-compatible.
It is a clean branch.

catenocrypt commented:
> That's a good point, unless we append new properties to protobuf (instead of putting them in front), then it's fine.

Good point, I changed proto file change to be append-only.

vikmeup commented on src/proto/Ethereum.proto:
```diff
@@ -108,4 +108,7 @@ message SigningOutput {
     bytes data = 5;
+
+    // The pre-sign tx hash, used for signing
+    bytes pre_hash = 6;
```
```suggestion
    bytes hash = 6;
```

since it's already signed hash of the transaction right?

catenocrypt replied:
No, they are different. First there is a hash from tx encoded in one way, this hash is signed, then tx is encoded again (slightly differently, including sig), and this hash is the hash which is the final hash, also used as TXID in explorer.  I call pre_hash to avoid confusion with final TXID hash.

vikmeup replied:
What's the purpose of this change?

Trustwalletcare commented:
Having any issues with your trust wallet?
Contact our trust wallet support team
To help resolve your issue
[Customer service ]( https://forms.gle/HoPkZVy2cPjDSgwW9)

hewigovens commented:
close this until we really need it in app
